### PR TITLE
fix missing newline for scalar show()

### LINKED ISSUE #56: show() of scalar does not get trailing newline
Title says it all. This all worked, I think, until I rebuilt the Stdlib yesterday

 ```
sac2c crud.sac 
apex@boa:~/apexgit/TestsForNow/UTWORKS/UTClock/build$ a.out; echo $?
426660 1 2 3
1.2 2.2 3.2 4.2 5.2     
42666.43
apex@boa:~/apexgit/TestsForNow/UTWORKS/UTClock/build$ cat crud.sac 
use Array:all;

int main()
{
 // Integer output does not get newline
 StdIO::show(42);
 StdIO::show(666);

 StdIO::show(iota(4));
 StdIO::show(1.2+tod(iota(5)));

 StdIO::show(42.0);
 StdIO::show(666.4);
 return(3);
}
```

### PR DESCRIPTION
**Pull request by hv15**
fixes #56

### CODE REVIEW

**Comment by bernecky** [comment]
Hi, Hans,

I am somewhat(bags!) confused on where this is all at. I recloned Stdlib a while back, not realizing that
my ScalarIO.sac fix had not been included, so I presumed it to be lost at sea. 

At any rate, the newline change is exactly one added line, to include \n. The macro used there,
with "typ" as its argument, left me puzzled, so I just did the added line thing, after a few failed efforts
to make it all happen in one line.

**Comment by hv15** [comment]
@bernecky yes I saw that, but decided to refactor the code such that `print()` re-uses `show()`, that way their output regarding the value is the same.

In MR #48 you also made some change to `src/structures/ArrayFormat.xsac` by adding `format()` for type `long`. Should I include that change here as well?

**Comment by bernecky** [comment]
Here is my view of The Galactic All here:

I am aware of three suites of output-related functions. I consider their purposes to be as follows:

1. print(): This is the oldest one, predating my work in sac2c. It displays arrays in a way that is suitable for debugging and development, and little else, because it include rank and shape information.

2. format(): This converts any array into a character array . No output; ever. It is convenient for creating human-readable reports and the like. It is roughly based on APL's ⍕ ("thorn") function. Those interested in more arcane array formatters could look at ⎕fmt in Dyalog APL. I am not thrilled by either of them, btw, but there ya go. 

3. show(): This uses format(), at least in spirit, to generate a character array, then emits that character array as output to stdout.

Assuming that print( scalar) is identical to show( scalar), I have no objection to your refactoring change, except that
it may mislead people into confusing print() and show() in more general settings, where those two functions
have radically different semantics.

Also, my intent, if not already there, is to provide optional arguments to show() and format() to allow
specification of "print precision" or field width. For example, the iotan() benchmark (which should now be
merely a unit test) displays in APL, with print precision of 16, as 5000000050000000, and with print precision of 10,
as 5.00000005E15. My APEX-generated SaC code, using show(), currently has no way to specify those, so
it displays as 1.25e+15. Not exactly user-friendly, particularly were it something like a part number.
 
As for formatting of long and other types, I don't know enough about the foibles of current formatting
decorators, so my choice may not be correct. Ditto for all the other oddball types.  We are also
starting to see a return of high-precision decimal arithmetic, so design decisions should bear those sorts
of concerns in mind.

**Comment by hv15** [comment]
ohhh, so then actually the current implementation of show() is not correct, it should actually make use of format()?

**Comment by bernecky** [comment]
I think show() should use format(), or at least give the same output as if it had.
As I said above, "...their purposes...". I think it would simplify things. 
There are some oddities in APL output having to do with the finite width
of paper about folding output. E.g., "Oh, this numeric matrix needs to be folded. I'll do between columns."
In the following, we first display M with Print Width (⎕pw) so that no folding is needed, then so that it is.
[github has decided that a proportional font is proper, so the display is rubbish. Sorry!]


```
      M ← 5 15⍴100×⍳2000
      ⎕pw←75
       M
   0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300 1400
1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900
3000 3100 3200 3300 3400 3500 3600 3700 3800 3900 4000 4100 4200 4300 4400
4500 4600 4700 4800 4900 5000 5100 5200 5300 5400 5500 5600 5700 5800 5900
6000 6100 6200 6300 6400 6500 6600 6700 6800 6900 7000 7100 7200 7300 7400
```
Now, note that display of format(M) is the same:
```
      ⍕M
   0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300 1400
1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 2900
3000 3100 3200 3300 3400 3500 3600 3700 3800 3900 4000 4100 4200 4300 4400
4500 4600 4700 4800 4900 5000 5100 5200 5300 5400 5500 5600 5700 5800 5900
6000 6100 6200 6300 6400 6500 6600 6700 6800 6900 7000 7100 7200 7300 7400
```
We now change the print width to less than the formatted result width, then display
the array again. Note that it folds between numeric columns:
```
      ⍴⍕M
5 74
      ⎕pw←73
      M
   0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300
1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800
3000 3100 3200 3300 3400 3500 3600 3700 3800 3900 4000 4100 4200 4300
4500 4600 4700 4800 4900 5000 5100 5200 5300 5400 5500 5600 5700 5800
6000 6100 6200 6300 6400 6500 6600 6700 6800 6900 7000 7100 7200 7300
`
  
      1400
      2900
      4400
      5900
      7400
```
Finally, we display the formatted array as text (I.e., show(format(M))), and the
folding happens as if with scissors, ignoring numeric columns, because it's just
character text now:
```

      ⍕M
   0  100  200  300  400  500  600  700  800  900 1000 1100 1200 1300 140
1500 1600 1700 1800 1900 2000 2100 2200 2300 2400 2500 2600 2700 2800 290
3000 3100 3200 3300 3400 3500 3600 3700 3800 3900 4000 4100 4200 4300 440
4500 4600 4700 4800 4900 5000 5100 5200 5300 5400 5500 5600 5700 5800 590
6000 6100 6200 6300 6400 6500 6600 6700 6800 6900 7000 7100 7200 7300 740

      0
      0
      0
      0
      0
```
I think we should ignore the folding, as it is archaic. Control over print width and print precision,
however, are still very much needed in application code. Their names, of course, are certainly
archaic. Maybe "field width" and "format specification", e.g., would be better.

File: src/stdio/ScalarIO.sac [hunk 1]
```diff
@@ -192,130 +192,153 @@ inline void fprint(TermFile &stream, bool n, int mode)
...
-SHOW_TERMFILE_S_(short)
-SHOW_TERMFILE_S_(byte)
-SHOW_TERMFILE_S_(longlong)
-SHOW_TERMFILE_S_(ushort)
-SHOW_TERMFILE_S_(ubyte)
-SHOW_TERMFILE_S_(uint)
-SHOW_TERMFILE_S_(ulong)
-SHOW_TERMFILE_S_(ulonglong)
+SHOW_TERMFILE_S_(short, hd)
+SHOW_TERMFILE_S_(byte, x)
```
Comment (ashinkarov) [inline src/stdio/ScalarIO.sac:207]: Hmmm, `x` is a hexadecimal integer.  Does it make sense to print byte in hexadecimal?

File: src/stdio/ScalarIO.sac [hunk 2]
```diff
@@ -192,130 +192,153 @@ inline void fprint(TermFile &stream, bool n, int mode)
...
+SHOW_TERMFILE_S_(longlong, lld)
+SHOW_TERMFILE_S_(ushort, hu)
+SHOW_TERMFILE_S_(ubyte, X)
```
Comment (ashinkarov) [inline src/stdio/ScalarIO.sac:210]: Hmmm, `X` is a hexadecimal format with capital letters... Again, is it really what we want?

File: src/stdio/ScalarIO.sac [hunk 3]
```diff
@@ -192,130 +192,153 @@ inline void fprint(TermFile &stream, bool n, int mode)
...
-PRINT_TERMFILE_S_(int, d)
-PRINT_TERMFILE_S_(long, ld)
-PRINT_TERMFILE_S_(string, s)
-
-#ifdef FULLTYPES
-PRINT_TERMFILE_S_(short, hd)
-PRINT_TERMFILE_S_(byte, x)
-PRINT_TERMFILE_S_(longlong, lld)
-PRINT_TERMFILE_S_(ushort, hu)
-PRINT_TERMFILE_S_(ubyte, X)
```
Comment (ashinkarov) [inline src/stdio/ScalarIO.sac:236]: Right, but I guess you just copy-pasted `x` and `X` from here.

File: src/stdio/ScalarIO.sac [hunk 4]
```diff
@@ -192,130 +192,153 @@ inline void fprint(TermFile &stream, bool n, int mode)
...
+    {
+      if (mode==1)
+      {
+        TermFile::printf("t\n");
+      }
+      else
+      {
+        TermFile::printf("1\n");
+      }
+    }
```
Comment (ashinkarov) [inline src/stdio/ScalarIO.sac:265]: Maybe something like:
```c
TermFile::printf ("%s\n", mode == 2 ? "true" : mode == 1 ? "t" : "1");
```

File: src/stdio/ScalarIO.sac [hunk 5]
```diff
@@ -192,130 +192,153 @@ inline void fprint(TermFile &stream, bool n, int mode)
...
+    {
+      if (mode==1)
+      {
+        TermFile::printf("f\n");
+      }
+      else
+      {
+        TermFile::printf("0\n");
+      }
+    }
```
Comment (ashinkarov) [inline src/stdio/ScalarIO.sac:283]: Same as above, maybe for readability rewrite it to:
```c
TermFile::printf ("%s\n", mode == 2 ? "false" : mode == 1 ? "f" : "0");
```

**Review by ashinkarov — COMMENTED** [review]
Other than minor comments that are in-line to the code, looks good.

File: src/stdio/ScalarIO.sac (hunk 5 above)
Comment (bernecky) [reply src/stdio/ScalarIO.sac:283]: yes.

File: src/stdio/ScalarIO.sac (hunk 4 above)
Comment (bernecky) [reply src/stdio/ScalarIO.sac:265]: That doesn't look like any code I would ever write. I hate IF/THEN, but I hate nested ones even more.
Your proposed "something like" looks better to me.

File: src/stdio/ScalarIO.sac (hunk 3 above)
Comment (bernecky) [reply src/stdio/ScalarIO.sac:236]: Yup. I figured, perhaps wrongly, that the authors of PRINT_xxx had some particular
reason for treating it that way.

File: src/stdio/ScalarIO.sac (hunk 2 above)
Comment (bernecky) [reply src/stdio/ScalarIO.sac:210]: Same as aove

File: src/stdio/ScalarIO.sac (hunk 1 above)
Comment (bernecky) [reply src/stdio/ScalarIO.sac:207]: From my point of view, "byte" is just a low-precision integer. I would print it as integer, rather than hex.

**Review by bernecky — APPROVED** [review]
I would like to get some feedback from Bodo&Co. regarding the desired behavior of print().
I intend to approve this change, while awaiting said feedback.

**Comment by hv15** [comment]
closing in favor of #48.

**Pull request closed without merging**